<a href="https://colab.research.google.com/github/Markwema86/-Investment-Database/blob/main/Investment_Database.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Setting Up SQLite In Colab

In [11]:
import sqlite3
import pandas as pd

# Creating our investment database in memory
conn = sqlite3.connect('investment_firm.db')
cursor = conn.cursor()

print("Database connected successfully.")
print("We are building: Naviel Capital Investment Database")

Database connected successfully.
We are building: Naviel Capital Investment Database


 #### Creating First Table: Companies

In [2]:
cursor.execute('''
CREATE TABLE IF NOT EXISTS companies (
    company_id      INTEGER PRIMARY KEY,
    company_name    TEXT NOT NULL,
    sector          TEXT,
    country         TEXT,
    founded_year    INTEGER,
    stage           TEXT,    -- Early, Growth, Mature
    status          TEXT     -- Portfolio, Exited, Prospect
)
''')

conn.commit()
print("Companies table created.")

Companies table created.


#### Creating Financials Table

In [3]:
cursor.execute('''
CREATE TABLE IF NOT EXISTS financials (
    financial_id    INTEGER PRIMARY KEY,
    company_id      INTEGER,
    fiscal_year     INTEGER,
    revenue         REAL,        -- in millions USD
    ebitda          REAL,        -- in millions USD
    net_income      REAL,
    total_debt      REAL,
    cash            REAL,
    employees       INTEGER,
    FOREIGN KEY (company_id) REFERENCES companies(company_id)
)
''')

conn.commit()
print("Financials table created.")

Financials table created.


####  Creating Investments Table

In [4]:
cursor.execute('''
CREATE TABLE IF NOT EXISTS investments (
    investment_id       INTEGER PRIMARY KEY,
    company_id          INTEGER,
    investment_date     TEXT,
    investment_amount   REAL,    -- millions USD invested
    equity_stake        REAL,    -- percentage owned
    entry_valuation     REAL,    -- company valuation at entry
    current_valuation   REAL,    -- current estimated valuation
    lead_partner        TEXT,
    FOREIGN KEY (company_id) REFERENCES companies(company_id)
)
''')

conn.commit()
print("Investments table created.")

Investments table created.


####  Creating SaaS Metrics Table

In [5]:
cursor.execute('''
CREATE TABLE IF NOT EXISTS saas_metrics (
    metric_id       INTEGER PRIMARY KEY,
    company_id      INTEGER,
    fiscal_year     INTEGER,
    fiscal_quarter  INTEGER,
    arr             REAL,    -- Annual Recurring Revenue (millions)
    mrr             REAL,    -- Monthly Recurring Revenue (millions)
    churn_rate      REAL,    -- percentage of customers lost
    nrr             REAL,    -- Net Revenue Retention (%)
    cac             REAL,    -- Customer Acquisition Cost (USD)
    ltv             REAL,    -- Customer Lifetime Value (USD)
    customers       INTEGER,
    FOREIGN KEY (company_id) REFERENCES companies(company_id)
)
''')

conn.commit()
print("SaaS metrics table created.")

SaaS metrics table created.


#### Populating With Realistic Data

In [6]:
# Insert companies
companies_data = [
    (1, 'TechFlow SaaS',      'Technology',   'Kenya',        2018, 'Growth',  'Portfolio'),
    (2, 'AfriPay Solutions',  'Fintech',       'Nigeria',      2019, 'Growth',  'Portfolio'),
    (3, 'HealthBridge',       'Healthcare',    'South Africa', 2016, 'Mature',  'Portfolio'),
    (4, 'LogiChain Africa',   'Logistics',     'Kenya',        2020, 'Early',   'Portfolio'),
    (5, 'SolarGrid Kenya',    'Clean Energy',  'Kenya',        2017, 'Growth',  'Portfolio'),
    (6, 'EduTech Nairobi',    'Education',     'Kenya',        2021, 'Early',   'Prospect'),
    (7, 'AgriData Systems',   'Agriculture',   'Ghana',        2015, 'Mature',  'Exited'),
    (8, 'CloudBase Africa',   'Technology',    'Rwanda',       2020, 'Growth',  'Portfolio'),
]

cursor.executemany('''
    INSERT OR REPLACE INTO companies
    VALUES (?,?,?,?,?,?,?)
''', companies_data)

# Insert financials
financials_data = [
    (1,  1, 2022,  4.2,  1.1,  0.4,  2.0,  1.5,  45),
    (2,  1, 2023,  6.8,  1.9,  0.9,  2.5,  2.1,  62),
    (3,  1, 2024,  9.4,  2.8,  1.4,  2.2,  3.8,  78),
    (4,  2, 2022,  8.1,  2.0,  0.8,  3.5,  2.2,  90),
    (5,  2, 2023, 11.3,  2.9,  1.2,  3.8,  3.1, 112),
    (6,  2, 2024, 15.7,  4.1,  2.1,  4.0,  5.4, 145),
    (7,  3, 2022, 22.4,  5.6,  2.8,  8.0,  4.1, 210),
    (8,  3, 2023, 24.1,  6.0,  3.1,  7.5,  5.2, 218),
    (9,  3, 2024, 26.8,  6.8,  3.6,  7.0,  6.8, 225),
    (10, 4, 2022,  1.8,  0.3, -0.2,  1.2,  0.8,  28),
    (11, 4, 2023,  3.2,  0.6,  0.1,  1.5,  1.1,  35),
    (12, 4, 2024,  5.1,  1.0,  0.4,  1.8,  1.9,  44),
    (13, 5, 2022,  6.5,  1.8,  0.9,  4.5,  1.2,  55),
    (14, 5, 2023,  8.2,  2.3,  1.2,  4.2,  2.0,  61),
    (15, 5, 2024, 10.9,  3.1,  1.8,  3.8,  3.2,  70),
    (16, 8, 2022,  2.9,  0.7,  0.1,  1.0,  0.9,  32),
    (17, 8, 2023,  4.8,  1.2,  0.4,  1.2,  1.8,  41),
    (18, 8, 2024,  7.3,  2.0,  0.9,  1.5,  3.1,  58),
]

cursor.executemany('''
    INSERT OR REPLACE INTO financials
    VALUES (?,?,?,?,?,?,?,?,?)
''', financials_data)

# Insert investments
investments_data = [
    (1, 1, '2021-03-15',  3.0, 25.0,  12.0,  38.0, 'Sarah Kimani'),
    (2, 2, '2020-11-20',  5.0, 20.0,  25.0,  78.0, 'James Oduya'),
    (3, 3, '2019-06-10',  8.0, 15.0,  53.0, 179.0, 'Sarah Kimani'),
    (4, 4, '2022-01-08',  2.0, 30.0,   6.7,  17.0, 'David Mwangi'),
    (5, 5, '2020-08-22',  4.0, 22.0,  18.0,  50.0, 'James Oduya'),
    (6, 8, '2021-09-14',  2.5, 28.0,   8.9,  26.0, 'David Mwangi'),
]

cursor.executemany('''
    INSERT OR REPLACE INTO investments
    VALUES (?,?,?,?,?,?,?,?)
''', investments_data)

# Insert SaaS metrics
saas_data = [
    (1,  1, 2023, 1,  3.8, 0.32, 4.2, 108, 850,  8500, None),
    (2,  1, 2023, 2,  4.1, 0.34, 3.8, 109, 820,  8900, None),
    (3,  1, 2023, 3,  4.6, 0.38, 3.5, 111, 800,  9200, None),
    (4,  1, 2023, 4,  5.2, 0.43, 3.1, 114, 780,  9800, None),
    (5,  1, 2024, 1,  6.1, 0.51, 2.8, 118, 760, 10500, None),
    (6,  1, 2024, 2,  7.0, 0.58, 2.5, 121, 740, 11200, None),
    (7,  1, 2024, 3,  7.8, 0.65, 2.3, 124, 720, 11800, None),
    (8,  1, 2024, 4,  8.9, 0.74, 2.0, 128, 700, 12500, None),
    (9,  2, 2023, 1,  7.2, 0.60, 5.8, 102, 1200, 6000, None),
    (10, 2, 2023, 2,  7.8, 0.65, 5.5, 103, 1150, 6200, None),
    (11, 2, 2023, 3,  8.5, 0.71, 5.1, 105, 1100, 6500, None),
    (12, 2, 2023, 4,  9.4, 0.78, 4.8, 107, 1050, 6900, None),
    (13, 2, 2024, 1, 10.6, 0.88, 4.2, 110, 1000, 7400, None),
    (14, 2, 2024, 2, 11.9, 0.99, 3.8, 113,  980, 7900, None),
    (15, 2, 2024, 3, 13.1, 1.09, 3.4, 116,  960, 8400, None),
    (16, 2, 2024, 4, 14.8, 1.23, 3.0, 119,  940, 9000, None),
]

cursor.executemany('''
    INSERT OR REPLACE INTO saas_metrics
    VALUES (?,?,?,?,?,?,?,?,?,?,?)
''', saas_data)

conn.commit()
print("All data inserted successfully.")
print("\nDatabase summary:")
for table in ['companies', 'financials', 'investments', 'saas_metrics']:
    count = cursor.execute(f"SELECT COUNT(*) FROM {table}").fetchone()[0]
    print(f"  {table}: {count} rows")

All data inserted successfully.

Database summary:
  companies: 8 rows
  financials: 18 rows
  investments: 6 rows
  saas_metrics: 16 rows


#### Finance Queries

In [7]:
# Helper function to run queries cleanly
def query(sql, description=""):
    if description:
        print(f"\n{'═'*50}")
        print(f"  {description}")
        print('═'*50)
    result = pd.read_sql_query(sql, conn)
    print(result.to_string(index=False))
    return result

In [8]:
# Q1: What companies are in our portfolio?
query('''
    SELECT
        company_name,
        sector,
        country,
        stage,
        status
    FROM companies
    WHERE status = 'Portfolio'
    ORDER BY sector
''', "PORTFOLIO COMPANIES")


══════════════════════════════════════════════════
  PORTFOLIO COMPANIES
══════════════════════════════════════════════════
     company_name       sector      country  stage    status
  SolarGrid Kenya Clean Energy        Kenya Growth Portfolio
AfriPay Solutions      Fintech      Nigeria Growth Portfolio
     HealthBridge   Healthcare South Africa Mature Portfolio
 LogiChain Africa    Logistics        Kenya  Early Portfolio
    TechFlow SaaS   Technology        Kenya Growth Portfolio
 CloudBase Africa   Technology       Rwanda Growth Portfolio


,company_name,sector,country,stage,status
0,SolarGrid Kenya,Clean Energy,Kenya,Growth,Portfolio
1,AfriPay Solutions,Fintech,Nigeria,Growth,Portfolio
2,HealthBridge,Healthcare,South Africa,Mature,Portfolio
3,LogiChain Africa,Logistics,Kenya,Early,Portfolio
4,TechFlow SaaS,Technology,Kenya,Growth,Portfolio
5,CloudBase Africa,Technology,Rwanda,Growth,Portfolio


In [9]:
# Q2: Which company grew revenue fastest in 2024?
query('''
    SELECT
        c.company_name,
        f.fiscal_year,
        f.revenue,
        f.ebitda,
        ROUND(f.ebitda / f.revenue * 100, 1) AS ebitda_margin_pct
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    WHERE f.fiscal_year = 2024
    ORDER BY f.revenue DESC
''', "2024 REVENUE RANKING WITH EBITDA MARGINS")


══════════════════════════════════════════════════
  2024 REVENUE RANKING WITH EBITDA MARGINS
══════════════════════════════════════════════════
     company_name  fiscal_year  revenue  ebitda  ebitda_margin_pct
     HealthBridge         2024     26.8     6.8               25.4
AfriPay Solutions         2024     15.7     4.1               26.1
  SolarGrid Kenya         2024     10.9     3.1               28.4
    TechFlow SaaS         2024      9.4     2.8               29.8
 CloudBase Africa         2024      7.3     2.0               27.4
 LogiChain Africa         2024      5.1     1.0               19.6


,company_name,fiscal_year,revenue,ebitda,ebitda_margin_pct
0,HealthBridge,2024,26.8,6.8,25.4
1,AfriPay Solutions,2024,15.7,4.1,26.1
2,SolarGrid Kenya,2024,10.9,3.1,28.4
3,TechFlow SaaS,2024,9.4,2.8,29.8
4,CloudBase Africa,2024,7.3,2.0,27.4
5,LogiChain Africa,2024,5.1,1.0,19.6


In [10]:
# Q3: What is our total invested capital and current portfolio value?
query('''
    SELECT
        c.company_name,
        i.investment_amount      AS invested_millions,
        i.equity_stake           AS stake_pct,
        i.entry_valuation        AS entry_val_millions,
        i.current_valuation      AS current_val_millions,
        ROUND(
            (i.current_valuation - i.entry_valuation)
            / i.entry_valuation * 100, 1
        )                        AS valuation_growth_pct,
        i.lead_partner
    FROM investments i
    JOIN companies c ON i.company_id = c.company_id
    ORDER BY valuation_growth_pct DESC
''', "PORTFOLIO VALUATION & RETURNS")


══════════════════════════════════════════════════
  PORTFOLIO VALUATION & RETURNS
══════════════════════════════════════════════════
     company_name  invested_millions  stake_pct  entry_val_millions  current_val_millions  valuation_growth_pct lead_partner
     HealthBridge                8.0       15.0                53.0                 179.0                 237.7 Sarah Kimani
    TechFlow SaaS                3.0       25.0                12.0                  38.0                 216.7 Sarah Kimani
AfriPay Solutions                5.0       20.0                25.0                  78.0                 212.0  James Oduya
 CloudBase Africa                2.5       28.0                 8.9                  26.0                 192.1 David Mwangi
  SolarGrid Kenya                4.0       22.0                18.0                  50.0                 177.8  James Oduya
 LogiChain Africa                2.0       30.0                 6.7                  17.0                 153.7 Dav

,company_name,invested_millions,stake_pct,entry_val_millions,current_val_millions,valuation_growth_pct,lead_partner
0,HealthBridge,8.0,15.0,53.0,179.0,237.7,Sarah Kimani
1,TechFlow SaaS,3.0,25.0,12.0,38.0,216.7,Sarah Kimani
2,AfriPay Solutions,5.0,20.0,25.0,78.0,212.0,James Oduya
3,CloudBase Africa,2.5,28.0,8.9,26.0,192.1,David Mwangi
4,SolarGrid Kenya,4.0,22.0,18.0,50.0,177.8,James Oduya
5,LogiChain Africa,2.0,30.0,6.7,17.0,153.7,David Mwangi


#### Analytics - make me reason
- How many portfolio companies do we have and which sectors are represented?

6 portfolio companies......sectors - Clean Energy, Fintech, Healthcare, Logistics, Technology

- Which company had the highest EBITDA margin in 2024 - and what does a high EBITDA margin tell a PE investor?

HealthBridge.....For a PE investor, a high EBITDA margin indicates that the company is financially efficient, capable of generating strong cash flows, and potentially scalable, making it a desirable candidate for leveraged acquisitions or growth-oriented investments. It also informs the valuation strategy and risk assessment, helping gauge both return potential and operational robustness.

- Which investment has grown the most in valuation — and by how much?

HealthBridge.....by 237.7%

- Looking at the database structure — why do you think we separated companies, financials, and investments into three different tables instead of one big table?

Separating tables eliminates data redundancy. If all data lived in one table, every time HealthBridge appeared you'd repeat its name, sector and country across every row of financial data. That wastes storage, creates inconsistency risk and makes updates painful. This principle is called database normalisation - and it's why every serious financial system is built this way.

# Aggregations, Grouping & Finance Questions

##### GROUP BY: Portfolio-Level Questions

In [12]:
# Q4: Total capital deployed by sector
query('''
    SELECT
        c.sector,
        COUNT(i.investment_id)              AS num_investments,
        ROUND(SUM(i.investment_amount), 1)  AS total_invested_millions,
        ROUND(AVG(i.equity_stake), 1)       AS avg_stake_pct,
        ROUND(SUM(i.current_valuation), 1)  AS total_current_value
    FROM investments i
    JOIN companies c ON i.company_id = c.company_id
    GROUP BY c.sector
    ORDER BY total_invested_millions DESC
''', "CAPITAL DEPLOYMENT BY SECTOR")


══════════════════════════════════════════════════
  CAPITAL DEPLOYMENT BY SECTOR
══════════════════════════════════════════════════
      sector  num_investments  total_invested_millions  avg_stake_pct  total_current_value
  Healthcare                1                      8.0           15.0                179.0
  Technology                2                      5.5           26.5                 64.0
     Fintech                1                      5.0           20.0                 78.0
Clean Energy                1                      4.0           22.0                 50.0
   Logistics                1                      2.0           30.0                 17.0


,sector,num_investments,total_invested_millions,avg_stake_pct,total_current_value
0,Healthcare,1,8.0,15.0,179.0
1,Technology,2,5.5,26.5,64.0
2,Fintech,1,5.0,20.0,78.0
3,Clean Energy,1,4.0,22.0,50.0
4,Logistics,1,2.0,30.0,17.0


In [13]:
# Q5: Revenue growth by company across all years
query('''
    SELECT
        c.company_name,
        MIN(f.revenue)  AS revenue_2022,
        MAX(f.revenue)  AS revenue_2024,
        ROUND(
            (MAX(f.revenue) - MIN(f.revenue))
            / MIN(f.revenue) * 100, 1
        )               AS total_growth_pct
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    GROUP BY c.company_name
    ORDER BY total_growth_pct DESC
''', "REVENUE GROWTH 2022-2024 BY COMPANY")


══════════════════════════════════════════════════
  REVENUE GROWTH 2022-2024 BY COMPANY
══════════════════════════════════════════════════
     company_name  revenue_2022  revenue_2024  total_growth_pct
 LogiChain Africa           1.8           5.1             183.3
 CloudBase Africa           2.9           7.3             151.7
    TechFlow SaaS           4.2           9.4             123.8
AfriPay Solutions           8.1          15.7              93.8
  SolarGrid Kenya           6.5          10.9              67.7
     HealthBridge          22.4          26.8              19.6


,company_name,revenue_2022,revenue_2024,total_growth_pct
0,LogiChain Africa,1.8,5.1,183.3
1,CloudBase Africa,2.9,7.3,151.7
2,TechFlow SaaS,4.2,9.4,123.8
3,AfriPay Solutions,8.1,15.7,93.8
4,SolarGrid Kenya,6.5,10.9,67.7
5,HealthBridge,22.4,26.8,19.6


In [16]:
# Q6: Which partner manages the most value?
query('''
    SELECT
        i.lead_partner,
        COUNT(i.investment_id)              AS companies_managed,
        ROUND(SUM(i.investment_amount), 1)  AS capital_deployed,
        ROUND(SUM(i.current_valuation), 1)  AS portfolio_value,
        ROUND(
            SUM(i.current_valuation) /
            SUM(i.investment_amount), 2
        )                                   AS value_multiple
    FROM investments i
    GROUP BY i.lead_partner
    ORDER BY portfolio_value DESC
''', "PARTNER PERFORMANCE - VALUE UNDER MANAGEMENT")


══════════════════════════════════════════════════
  PARTNER PERFORMANCE - VALUE UNDER MANAGEMENT
══════════════════════════════════════════════════
lead_partner  companies_managed  capital_deployed  portfolio_value  value_multiple
Sarah Kimani                  2              11.0            217.0           19.73
 James Oduya                  2               9.0            128.0           14.22
David Mwangi                  2               4.5             43.0            9.56


,lead_partner,companies_managed,capital_deployed,portfolio_value,value_multiple
0,Sarah Kimani,2,11.0,217.0,19.73
1,James Oduya,2,9.0,128.0,14.22
2,David Mwangi,2,4.5,43.0,9.56


##### HAVING: Filtering After Grouping
**WHERE** filters rows before grouping. **HAVING** filters after. Critical distinction.

In [17]:
# Q7: Which companies consistently grew revenue every year?
# Only show companies where ALL years show positive EBITDA
query('''
    SELECT
        c.company_name,
        COUNT(f.fiscal_year)        AS years_tracked,
        MIN(f.ebitda)               AS lowest_ebitda,
        MAX(f.ebitda)               AS highest_ebitda,
        ROUND(AVG(f.ebitda), 2)     AS avg_ebitda
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    GROUP BY c.company_name
    HAVING MIN(f.ebitda) > 0
    ORDER BY avg_ebitda DESC
''', "CONSISTENTLY PROFITABLE COMPANIES (All years EBITDA > 0)")


══════════════════════════════════════════════════
  CONSISTENTLY PROFITABLE COMPANIES (All years EBITDA > 0)
══════════════════════════════════════════════════
     company_name  years_tracked  lowest_ebitda  highest_ebitda  avg_ebitda
     HealthBridge              3            5.6             6.8        6.13
AfriPay Solutions              3            2.0             4.1        3.00
  SolarGrid Kenya              3            1.8             3.1        2.40
    TechFlow SaaS              3            1.1             2.8        1.93
 CloudBase Africa              3            0.7             2.0        1.30
 LogiChain Africa              3            0.3             1.0        0.63


,company_name,years_tracked,lowest_ebitda,highest_ebitda,avg_ebitda
0,HealthBridge,3,5.6,6.8,6.13
1,AfriPay Solutions,3,2.0,4.1,3.00
2,SolarGrid Kenya,3,1.8,3.1,2.40
3,TechFlow SaaS,3,1.1,2.8,1.93
4,CloudBase Africa,3,0.7,2.0,1.30
5,LogiChain Africa,3,0.3,1.0,0.63


In [18]:
# Q8: Flag high-debt companies - PE due diligence risk screen
query('''
    SELECT
        c.company_name,
        f.fiscal_year,
        f.total_debt,
        f.ebitda,
        ROUND(f.total_debt / f.ebitda, 2)   AS debt_to_ebitda,
        CASE
            WHEN f.total_debt / f.ebitda > 5 THEN "HIGH RISK"
            WHEN f.total_debt / f.ebitda > 3 THEN "MONITOR"
            ELSE "HEALTHY"
        END                                  AS debt_risk_flag
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    WHERE f.fiscal_year = 2024
    ORDER BY debt_to_ebitda DESC
''', "DEBT RISK SCREENING — 2024")


══════════════════════════════════════════════════
  DEBT RISK SCREENING — 2024
══════════════════════════════════════════════════
     company_name  fiscal_year  total_debt  ebitda  debt_to_ebitda debt_risk_flag
 LogiChain Africa         2024         1.8     1.0            1.80        HEALTHY
  SolarGrid Kenya         2024         3.8     3.1            1.23        HEALTHY
     HealthBridge         2024         7.0     6.8            1.03        HEALTHY
AfriPay Solutions         2024         4.0     4.1            0.98        HEALTHY
    TechFlow SaaS         2024         2.2     2.8            0.79        HEALTHY
 CloudBase Africa         2024         1.5     2.0            0.75        HEALTHY


,company_name,fiscal_year,total_debt,ebitda,debt_to_ebitda,debt_risk_flag
0,LogiChain Africa,2024,1.8,1.0,1.80,HEALTHY
1,SolarGrid Kenya,2024,3.8,3.1,1.23,HEALTHY
2,HealthBridge,2024,7.0,6.8,1.03,HEALTHY
3,AfriPay Solutions,2024,4.0,4.1,0.98,HEALTHY
4,TechFlow SaaS,2024,2.2,2.8,0.79,HEALTHY
5,CloudBase Africa,2024,1.5,2.0,0.75,HEALTHY


**Why this matters:** Debt-to-EBITDA above 4 - 5x is a red flag in PE. Lenders typically won't finance an acquisition if this ratio is too high.

###### CASE WHEN: Business Logic In SQL

In [20]:
# Q9: Investment performance classification
query('''
    SELECT
        c.company_name,
        i.investment_amount,
        i.current_valuation,
        ROUND(i.current_valuation / i.investment_amount, 2) AS moic,
        CASE
            WHEN i.current_valuation / i.investment_amount >= 3.0
                THEN "STAR - Exceptional (3x+)"
            WHEN i.current_valuation / i.investment_amount >= 2.0
                THEN "STRONG - Good Return (2-3x)"
            WHEN i.current_valuation / i.investment_amount >= 1.5
                THEN "PERFORMING - On Track (1.5-2x)"
            WHEN i.current_valuation / i.investment_amount >= 1.0
                THEN "WATCH - Below Target (1-1.5x)"
            ELSE
                "UNDERPERFORMING - Capital at Risk (<1x)"
        END AS performance_band
    FROM investments i
    JOIN companies c ON i.company_id = c.company_id
    ORDER BY moic DESC
''', "INVESTMENT PERFORMANCE - MOIC CLASSIFICATION")


══════════════════════════════════════════════════
  INVESTMENT PERFORMANCE - MOIC CLASSIFICATION
══════════════════════════════════════════════════
     company_name  investment_amount  current_valuation  moic         performance_band
     HealthBridge                8.0              179.0 22.38 STAR - Exceptional (3x+)
AfriPay Solutions                5.0               78.0 15.60 STAR - Exceptional (3x+)
    TechFlow SaaS                3.0               38.0 12.67 STAR - Exceptional (3x+)
  SolarGrid Kenya                4.0               50.0 12.50 STAR - Exceptional (3x+)
 CloudBase Africa                2.5               26.0 10.40 STAR - Exceptional (3x+)
 LogiChain Africa                2.0               17.0  8.50 STAR - Exceptional (3x+)


,company_name,investment_amount,current_valuation,moic,performance_band
0,HealthBridge,8.0,179.0,22.38,STAR - Exceptional (3x+)
1,AfriPay Solutions,5.0,78.0,15.60,STAR - Exceptional (3x+)
2,TechFlow SaaS,3.0,38.0,12.67,STAR - Exceptional (3x+)
3,SolarGrid Kenya,4.0,50.0,12.50,STAR - Exceptional (3x+)
4,CloudBase Africa,2.5,26.0,10.40,STAR - Exceptional (3x+)
5,LogiChain Africa,2.0,17.0,8.50,STAR - Exceptional (3x+)


**MOIC** = **Multiple on Invested Capital**. It's the primary return metric in PE. A 3x MOIC means every $1

invested returned $3. Top PE funds target 3x+ on their best deals.

#### Combine Everything Into A Portfolio Dashboard Query - Executive

In [22]:
# Q10: Full portfolio health dashboard - one query
query('''
    SELECT
        c.company_name,
        c.sector,
        c.country,
        f.revenue                                           AS revenue_2024,
        ROUND(f.ebitda / f.revenue * 100, 1)               AS ebitda_margin,
        ROUND(f.total_debt / f.ebitda, 2)                  AS debt_to_ebitda,
        ROUND(i.current_valuation / i.investment_amount, 2) AS moic,
        i.lead_partner,
        CASE
            WHEN f.ebitda / f.revenue > 0.25
             AND f.total_debt / f.ebitda < 3
             AND i.current_valuation / i.investment_amount >= 2
                THEN "GREEN - Strong"
            WHEN f.ebitda / f.revenue > 0.15
             AND f.total_debt / f.ebitda < 4
                THEN "AMBER - Monitor"
            ELSE
                "RED - Action Required"
        END AS portfolio_status
    FROM financials f
    JOIN companies c  ON f.company_id  = c.company_id
    JOIN investments i ON c.company_id = i.company_id
    WHERE f.fiscal_year = 2024
      AND c.status = "Portfolio"
    ORDER BY moic DESC
''', "PORTFOLIO HEALTH DASHBOARD - Q4 2024")


══════════════════════════════════════════════════
  PORTFOLIO HEALTH DASHBOARD - Q4 2024
══════════════════════════════════════════════════
     company_name       sector      country  revenue_2024  ebitda_margin  debt_to_ebitda  moic lead_partner portfolio_status
     HealthBridge   Healthcare South Africa          26.8           25.4            1.03 22.38 Sarah Kimani   GREEN - Strong
AfriPay Solutions      Fintech      Nigeria          15.7           26.1            0.98 15.60  James Oduya   GREEN - Strong
    TechFlow SaaS   Technology        Kenya           9.4           29.8            0.79 12.67 Sarah Kimani   GREEN - Strong
  SolarGrid Kenya Clean Energy        Kenya          10.9           28.4            1.23 12.50  James Oduya   GREEN - Strong
 CloudBase Africa   Technology       Rwanda           7.3           27.4            0.75 10.40 David Mwangi   GREEN - Strong
 LogiChain Africa    Logistics        Kenya           5.1           19.6            1.80  8.50 David Mwangi 

,company_name,sector,country,revenue_2024,ebitda_margin,debt_to_ebitda,moic,lead_partner,portfolio_status
0,HealthBridge,Healthcare,South Africa,26.8,25.4,1.03,22.38,Sarah Kimani,GREEN - Strong
1,AfriPay Solutions,Fintech,Nigeria,15.7,26.1,0.98,15.60,James Oduya,GREEN - Strong
2,TechFlow SaaS,Technology,Kenya,9.4,29.8,0.79,12.67,Sarah Kimani,GREEN - Strong
3,SolarGrid Kenya,Clean Energy,Kenya,10.9,28.4,1.23,12.50,James Oduya,GREEN - Strong
4,CloudBase Africa,Technology,Rwanda,7.3,27.4,0.75,10.40,David Mwangi,GREEN - Strong
5,LogiChain Africa,Logistics,Kenya,5.1,19.6,1.80,8.50,David Mwangi,AMBER - Monitor


#### Analysis - make me reason
- Which partner has the highest value multiple  and what does that tell you about their investment selection?

Sarah Kimani has the highest value multiple of 19.73.

This indicates that for every dollar Sarah Kimani invested, it returned $19.73 in current portfolio value. A high value multiple like this suggests exceptional investment selection, effective management of the portfolio companies, and successful growth strategies for the assets under her management.

- Which companies got flagged as debt risk - and what would you do as a portfolio manager?

no companies were flagged as having high debt risk ('HIGH RISK' or 'MONITOR') for the fiscal year 2024. All companies currently show a 'HEALTHY' debt-to-EBITDA ratio.

However, if any companies were flagged as 'HIGH RISK' (debt-to-EBITDA > 5) or 'MONITOR' (debt-to-EBITDA > 3), a portfolio manager would typically take the following actions:

**For 'MONITOR' companies:**

**Deeper Dive:** Conduct a more in-depth financial analysis to understand the reasons behind the elevated debt. This might involve reviewing cash flow statements, upcoming debt maturities, and capital expenditure plans.

**Management Discussions:** Engage with the company's management team to discuss their strategy for debt reduction, operational improvements, and growth plans.

**Scenario Planning:** Model different financial scenarios to assess the company's resilience under various market conditions.

**Reporting:** Increase the frequency of financial reporting and key performance indicator (KPI) tracking for these companies.

**For 'HIGH RISK' companies:**

**Immediate Action Plan:** Work with management to develop and implement an urgent plan to reduce debt or improve EBITDA. This could include cost-cutting measures, divestitures of non-core assets, or strategic equity raises.

**Covenant Review:** Scrutinize all debt covenants to ensure no breaches are imminent, and proactively engage with lenders if potential issues arise.

**Contingency Planning:** Develop contingency plans, including potential restructuring options or, in extreme cases, exit strategies.

**Increased Oversight:** Implement more rigorous oversight and potentially assign additional resources (e.g., operational experts) to support the company.

In both cases, the goal is to protect the firm's investment and guide the portfolio company back to a healthier financial position.

- Which investments hit STAR classification (3x MOIC) - and which are underperforming?

Investments that hit STAR classification (3x MOIC):

HealthBridge (22.38x MOIC)

AfriPay Solutions (15.60x MOIC)

TechFlow SaaS (12.67x MOIC)

SolarGrid Kenya (12.50x MOIC)

CloudBase Africa (10.40x MOIC)

LogiChain Africa (8.50x MOIC)

Underperforming investments (Capital at Risk < 1x):

There are no companies currently classified as 'UNDERPERFORMING - Capital at Risk (< 1x)' in the portfolio. All investments are showing an MOIC significantly above 1x, indicating strong performance across the board.

- In the portfolio dashboard - how many companies are GREEN, AMBER, and RED?

**GREEN - Strong: 5 companies**

HealthBridge

AfriPay Solutions

TechFlow SaaS

SolarGrid Kenya

CloudBase Africa

**AMBER - Monitor: 1 company**

LogiChain Africa

**RED - Action Required: 0 companies**

#### new SQL concepts used
- GROUP BY - aggregating rows into summary statistics
- HAVING - filtering after aggregation (different from WHERE)
- CASE WHEN - business logic and classification inside SQL
- Multi-table JOINs - connecting three tables simultaneously
- Calculated columns - MOIC, debt-to-EBITDA, EBITDA margin computed inside the query

# Window Functions
most powerful tool in financial SQL - and the most commonly tested in interviews

**What they do:** Calculate values across rows without collapsing them into groups. You keep every row AND get aggregate context alongside it.

**GROUP BY asks:** "What is the average revenue across all companies?"

**Window function asks:** "What is the average revenue - AND show me each company's revenue next to that average on the same row?"

Window Function: Ranking

In [23]:
# Q11: Rank companies by revenue within each sector
query('''
    SELECT
        c.company_name,
        c.sector,
        f.fiscal_year,
        f.revenue,
        RANK() OVER (
            PARTITION BY c.sector
            ORDER BY f.revenue DESC
        ) AS sector_revenue_rank,
        RANK() OVER (
            ORDER BY f.revenue DESC
        ) AS overall_revenue_rank
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    WHERE f.fiscal_year = 2024
    ORDER BY c.sector, sector_revenue_rank
''', "REVENUE RANKING - WITHIN SECTOR AND OVERALL (2024)")


══════════════════════════════════════════════════
  REVENUE RANKING — WITHIN SECTOR AND OVERALL (2024)
══════════════════════════════════════════════════
     company_name       sector  fiscal_year  revenue  sector_revenue_rank  overall_revenue_rank
  SolarGrid Kenya Clean Energy         2024     10.9                    1                     3
AfriPay Solutions      Fintech         2024     15.7                    1                     2
     HealthBridge   Healthcare         2024     26.8                    1                     1
 LogiChain Africa    Logistics         2024      5.1                    1                     6
    TechFlow SaaS   Technology         2024      9.4                    1                     4
 CloudBase Africa   Technology         2024      7.3                    2                     5


,company_name,sector,fiscal_year,revenue,sector_revenue_rank,overall_revenue_rank
0,SolarGrid Kenya,Clean Energy,2024,10.9,1,3
1,AfriPay Solutions,Fintech,2024,15.7,1,2
2,HealthBridge,Healthcare,2024,26.8,1,1
3,LogiChain Africa,Logistics,2024,5.1,1,6
4,TechFlow SaaS,Technology,2024,9.4,1,4
5,CloudBase Africa,Technology,2024,7.3,2,5


Breaking down the syntax:

- RANK() - assigns a rank number
- OVER() - tells SQL this is a window function
- PARTITION BY sector - restart ranking for each sector
- ORDER BY revenue DESC - rank by revenue, highest first

Running Totals: Tracking Cash Deployment Over Time

In [24]:
# Q12: Cumulative capital deployed over time
query('''
    SELECT
        c.company_name,
        i.investment_date,
        i.investment_amount,
        SUM(i.investment_amount) OVER (
            ORDER BY i.investment_date
        ) AS cumulative_deployed,
        COUNT(i.investment_id) OVER (
            ORDER BY i.investment_date
        ) AS deals_completed
    FROM investments i
    JOIN companies c ON i.company_id = c.company_id
    ORDER BY i.investment_date
''', "CUMULATIVE CAPITAL DEPLOYMENT OVER TIME")


══════════════════════════════════════════════════
  CUMULATIVE CAPITAL DEPLOYMENT OVER TIME
══════════════════════════════════════════════════
     company_name investment_date  investment_amount  cumulative_deployed  deals_completed
     HealthBridge      2019-06-10                8.0                  8.0                1
  SolarGrid Kenya      2020-08-22                4.0                 12.0                2
AfriPay Solutions      2020-11-20                5.0                 17.0                3
    TechFlow SaaS      2021-03-15                3.0                 20.0                4
 CloudBase Africa      2021-09-14                2.5                 22.5                5
 LogiChain Africa      2022-01-08                2.0                 24.5                6


,company_name,investment_date,investment_amount,cumulative_deployed,deals_completed
0,HealthBridge,2019-06-10,8.0,8.0,1
1,SolarGrid Kenya,2020-08-22,4.0,12.0,2
2,AfriPay Solutions,2020-11-20,5.0,17.0,3
3,TechFlow SaaS,2021-03-15,3.0,20.0,4
4,CloudBase Africa,2021-09-14,2.5,22.5,5
5,LogiChain Africa,2022-01-08,2.0,24.5,6


This is the running total pattern - used in every fund reporting system to show how capital has been deployed over a fund's life.

**LAG & LEAD:** Year-Over-Year Growth

This is the most used window function in finance. **LAG** looks backward. **LEAD** looks forward.

In [25]:
# Q13: Year-over-year revenue growth per company
query('''
    SELECT
        c.company_name,
        f.fiscal_year,
        f.revenue,
        LAG(f.revenue) OVER (
            PARTITION BY f.company_id
            ORDER BY f.fiscal_year
        ) AS prior_year_revenue,
        ROUND(
            (f.revenue - LAG(f.revenue) OVER (
                PARTITION BY f.company_id
                ORDER BY f.fiscal_year
            )) /
            LAG(f.revenue) OVER (
                PARTITION BY f.company_id
                ORDER BY f.fiscal_year
            ) * 100, 1
        ) AS yoy_growth_pct
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    ORDER BY c.company_name, f.fiscal_year
''', "YEAR-OVER-YEAR REVENUE GROWTH PER COMPANY")


══════════════════════════════════════════════════
  YEAR-OVER-YEAR REVENUE GROWTH PER COMPANY
══════════════════════════════════════════════════
     company_name  fiscal_year  revenue  prior_year_revenue  yoy_growth_pct
AfriPay Solutions         2022      8.1                 NaN             NaN
AfriPay Solutions         2023     11.3                 8.1            39.5
AfriPay Solutions         2024     15.7                11.3            38.9
 CloudBase Africa         2022      2.9                 NaN             NaN
 CloudBase Africa         2023      4.8                 2.9            65.5
 CloudBase Africa         2024      7.3                 4.8            52.1
     HealthBridge         2022     22.4                 NaN             NaN
     HealthBridge         2023     24.1                22.4             7.6
     HealthBridge         2024     26.8                24.1            11.2
 LogiChain Africa         2022      1.8                 NaN             NaN
 LogiChain Africa

,company_name,fiscal_year,revenue,prior_year_revenue,yoy_growth_pct
0,AfriPay Solutions,2022,8.1,NaN,NaN
1,AfriPay Solutions,2023,11.3,8.1,39.5
2,AfriPay Solutions,2024,15.7,11.3,38.9
3,CloudBase Africa,2022,2.9,NaN,NaN
4,CloudBase Africa,2023,4.8,2.9,65.5
5,CloudBase Africa,2024,7.3,4.8,52.1
6,HealthBridge,2022,22.4,NaN,NaN
7,HealthBridge,2023,24.1,22.4,7.6
8,HealthBridge,2024,26.8,24.1,11.2
9,LogiChain Africa,2022,1.8,NaN,NaN


What LAG does: For each row, it looks back one row (within the same company, ordered by year) and returns that value. The first year shows NULL because there's no prior year.

 Percentile Rankings: Where Does Each Company Stand?

In [26]:
# Q14: EBITDA margin percentile across portfolio
query('''
    SELECT
        c.company_name,
        f.fiscal_year,
        ROUND(f.ebitda / f.revenue * 100, 1)    AS ebitda_margin_pct,
        ROUND(PERCENT_RANK() OVER (
            PARTITION BY f.fiscal_year
            ORDER BY f.ebitda / f.revenue
        ) * 100, 0)                              AS percentile_rank,
        NTILE(4) OVER (
            PARTITION BY f.fiscal_year
            ORDER BY f.ebitda / f.revenue DESC
        )                                        AS quartile
    FROM financials f
    JOIN companies c ON f.company_id = c.company_id
    WHERE f.fiscal_year = 2024
    ORDER BY ebitda_margin_pct DESC
''', "EBITDA MARGIN PERCENTILE RANKING (2024)")


══════════════════════════════════════════════════
  EBITDA MARGIN PERCENTILE RANKING (2024)
══════════════════════════════════════════════════
     company_name  fiscal_year  ebitda_margin_pct  percentile_rank  quartile
    TechFlow SaaS         2024               29.8            100.0         1
  SolarGrid Kenya         2024               28.4             80.0         1
 CloudBase Africa         2024               27.4             60.0         2
AfriPay Solutions         2024               26.1             40.0         2
     HealthBridge         2024               25.4             20.0         3
 LogiChain Africa         2024               19.6              0.0         4


,company_name,fiscal_year,ebitda_margin_pct,percentile_rank,quartile
0,TechFlow SaaS,2024,29.8,100.0,1
1,SolarGrid Kenya,2024,28.4,80.0,1
2,CloudBase Africa,2024,27.4,60.0,2
3,AfriPay Solutions,2024,26.1,40.0,2
4,HealthBridge,2024,25.4,20.0,3
5,LogiChain Africa,2024,19.6,0.0,4


NTILE(4) divides companies into quartiles - Q1 is top 25%, Q4 is bottom 25%. PE firms use this constantly to benchmark portfolio companies against each other.

Moving Averages: SaaS Metrics Trend Analysis

This connects SQL directly back to my earlier Python work - but now done in a database.

In [27]:
# Q15: Quarterly ARR with 4-quarter moving average
query('''
    SELECT
        c.company_name,
        s.fiscal_year,
        s.fiscal_quarter,
        s.arr,
        s.churn_rate,
        ROUND(AVG(s.arr) OVER (
            PARTITION BY s.company_id
            ORDER BY s.fiscal_year, s.fiscal_quarter
            ROWS BETWEEN 3 PRECEDING AND CURRENT ROW
        ), 2) AS arr_4q_moving_avg,
        ROUND(AVG(s.churn_rate) OVER (
            PARTITION BY s.company_id
            ORDER BY s.fiscal_year, s.fiscal_quarter
            ROWS BETWEEN 3 PRECEDING AND CURRENT ROW
        ), 2) AS churn_4q_moving_avg
    FROM saas_metrics s
    JOIN companies c ON s.company_id = c.company_id
    ORDER BY c.company_name, s.fiscal_year, s.fiscal_quarter
''', "QUARTERLY ARR & CHURN WITH 4-QUARTER MOVING AVERAGES")


══════════════════════════════════════════════════
  QUARTERLY ARR & CHURN WITH 4-QUARTER MOVING AVERAGES
══════════════════════════════════════════════════
     company_name  fiscal_year  fiscal_quarter  arr  churn_rate  arr_4q_moving_avg  churn_4q_moving_avg
AfriPay Solutions         2023               1  7.2         5.8               7.20                 5.80
AfriPay Solutions         2023               2  7.8         5.5               7.50                 5.65
AfriPay Solutions         2023               3  8.5         5.1               7.83                 5.47
AfriPay Solutions         2023               4  9.4         4.8               8.23                 5.30
AfriPay Solutions         2024               1 10.6         4.2               9.08                 4.90
AfriPay Solutions         2024               2 11.9         3.8              10.10                 4.48
AfriPay Solutions         2024               3 13.1         3.4              11.25                 4.05
AfriPay So

,company_name,fiscal_year,fiscal_quarter,arr,churn_rate,arr_4q_moving_avg,churn_4q_moving_avg
0,AfriPay Solutions,2023,1,7.2,5.8,7.20,5.80
1,AfriPay Solutions,2023,2,7.8,5.5,7.50,5.65
2,AfriPay Solutions,2023,3,8.5,5.1,7.83,5.47
3,AfriPay Solutions,2023,4,9.4,4.8,8.23,5.30
4,AfriPay Solutions,2024,1,10.6,4.2,9.08,4.90
5,AfriPay Solutions,2024,2,11.9,3.8,10.10,4.48
6,AfriPay Solutions,2024,3,13.1,3.4,11.25,4.05
7,AfriPay Solutions,2024,4,14.8,3.0,12.60,3.60
8,TechFlow SaaS,2023,1,3.8,4.2,3.80,4.20
9,TechFlow SaaS,2023,2,4.1,3.8,3.95,4.00


ROWS BETWEEN 3 PRECEDING AND CURRENT ROW - this defines the window frame. It says: look at this row and the 3 rows before it. That gives you a rolling 4-quarter average - identical to the rolling means I built in Python, but done entirely in SQL.

Full Window Function Dashboard

In [29]:
# Q16: Complete SaaS health scorecard using window functions
query('''
    SELECT
        c.company_name,
        s.fiscal_year,
        s.fiscal_quarter,
        s.arr,
        s.churn_rate,
        s.nrr,
        ROUND(s.ltv / s.cac, 2)             AS ltv_cac_ratio,
        RANK() OVER (
            PARTITION BY s.fiscal_year, s.fiscal_quarter
            ORDER BY s.nrr DESC
        )                                    AS nrr_rank,
        ROUND(s.arr - LAG(s.arr) OVER (
            PARTITION BY s.company_id
            ORDER BY s.fiscal_year, s.fiscal_quarter
        ), 2)                                AS arr_added_vs_prior_q,
        CASE
            WHEN s.nrr >= 120 THEN "ELITE - Best in class"
            WHEN s.nrr >= 110 THEN "STRONG - Above average"
            WHEN s.nrr >= 100 THEN "HEALTHY - Retaining revenue"
            ELSE                   "AT RISK - Losing revenue"
        END                                  AS nrr_classification
    FROM saas_metrics s
    JOIN companies c ON s.company_id = c.company_id
    ORDER BY c.company_name, s.fiscal_year, s.fiscal_quarter
''', "SAAS HEALTH SCORECARD WITH WINDOW FUNCTIONS")


══════════════════════════════════════════════════
  SAAS HEALTH SCORECARD WITH WINDOW FUNCTIONS
══════════════════════════════════════════════════
     company_name  fiscal_year  fiscal_quarter  arr  churn_rate   nrr  ltv_cac_ratio  nrr_rank  arr_added_vs_prior_q          nrr_classification
AfriPay Solutions         2023               1  7.2         5.8 102.0           5.00         2                   NaN HEALTHY - Retaining revenue
AfriPay Solutions         2023               2  7.8         5.5 103.0           5.39         2                   0.6 HEALTHY - Retaining revenue
AfriPay Solutions         2023               3  8.5         5.1 105.0           5.91         2                   0.7 HEALTHY - Retaining revenue
AfriPay Solutions         2023               4  9.4         4.8 107.0           6.57         2                   0.9 HEALTHY - Retaining revenue
AfriPay Solutions         2024               1 10.6         4.2 110.0           7.40         2                   1.2      STRO

,company_name,fiscal_year,fiscal_quarter,arr,churn_rate,nrr,ltv_cac_ratio,nrr_rank,arr_added_vs_prior_q,nrr_classification
0,AfriPay Solutions,2023,1,7.2,5.8,102.0,5.00,2,NaN,HEALTHY - Retaining revenue
1,AfriPay Solutions,2023,2,7.8,5.5,103.0,5.39,2,0.6,HEALTHY - Retaining revenue
2,AfriPay Solutions,2023,3,8.5,5.1,105.0,5.91,2,0.7,HEALTHY - Retaining revenue
3,AfriPay Solutions,2023,4,9.4,4.8,107.0,6.57,2,0.9,HEALTHY - Retaining revenue
4,AfriPay Solutions,2024,1,10.6,4.2,110.0,7.40,2,1.2,STRONG - Above average
5,AfriPay Solutions,2024,2,11.9,3.8,113.0,8.06,2,1.3,STRONG - Above average
6,AfriPay Solutions,2024,3,13.1,3.4,116.0,8.75,2,1.2,STRONG - Above average
7,AfriPay Solutions,2024,4,14.8,3.0,119.0,9.57,2,1.7,STRONG - Above average
8,TechFlow SaaS,2023,1,3.8,4.2,108.0,10.00,1,NaN,HEALTHY - Retaining revenue
9,TechFlow SaaS,2023,2,4.1,3.8,109.0,10.85,1,0.3,HEALTHY - Retaining revenue


**LTV/CAC ratio** - if it costs $ 1,000

 to acquire a customer (CAC) and they generate $10,000 over their lifetime (LTV), the ratio is 10x. Top SaaS companies target 3x minimum. Below 1x means you're losing money on every customer.

**NRR (Net Revenue Retention)** above 100% means existing customers are spending MORE over time - the holy grail of SaaS. It means the company grows even without acquiring new customers.